docs/man/httxt2dbm.1
.\" XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
.\" DO NOT EDIT! Generated from XML source.
.\" XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "HTTXT2DBM" 1 "2026-04-30" "Apache HTTP Server" "httxt2dbm"

.SH NAME
httxt2dbm \- Generate dbm files for use with RewriteMap

.SH "SYNOPSIS"
 
.PP
\fB\fBhttxt2dbm\fR [ -\fBv\fR ] [ -\fBf\fR \fIDBM_TYPE\fR ] -\fBi\fR \fISOURCE_TXT\fR -\fBo\fR \fIOUTPUT_DBM\fR \fR
 

.SH "SUMMARY"
 
.PP
\fBhttxt2dbm\fR is used to generate dbm files from text input, for use in RewriteMap with the \fBdbm\fR map type\&.
 
.PP
If the output file already exists, it will not be truncated\&. New keys will be added and existing keys will be updated\&.
 

.SH "OPTIONS"
 
 
.TP
\fB-v\fR
More verbose output  
.TP
\fB-f \fIDBM_TYPE\fR\fR
Specify the DBM type to be used for the output\&. If not specified, will use the APR Default\&. Available types are: \fBGDBM\fR for GDBM files, \fBSDBM\fR for SDBM files, \fBDB\fR for berkeley DB files, \fBNDBM\fR for NDBM files, \fBdefault\fR for the default DBM type\&.  
.TP
\fB-i \fISOURCE_TXT\fR\fR
Input file from which the dbm is to be created\&. The file should be formatted with one record per line, of the form: \fBkey value\fR\&. See the documentation for RewriteMap for further details of this file's format and meaning\&.  
.TP
\fB-o \fIOUTPUT_DBM\fR\fR
Name of the output dbm files\&.  
 
.SH "EXAMPLES"
 
.nf
 
httxt2dbm -i rewritemap\&.txt -o rewritemap\&.dbm
httxt2dbm -f SDBM -i rewritemap\&.txt -o rewritemap\&.dbm
 
.fi
 


CHANGES
                                                         -*- coding: utf-8 -*-
Changes with Apache 2.4.68

Changes with Apache 2.4.67

  *) SECURITY: CVE-2026-34059: Apache HTTP Server: mod_proxy_ajp:
     Heap Over-Read and memory disclosure in  ajp_parse_data()
     (cve.mitre.org)
     Buffer Over-read vulnerability in Apache HTTP Server.
     This issue affects Apache HTTP Server: through 2.4.66.
     Credits: Elhanan Haenel

  *) SECURITY: CVE-2026-34032: Apache HTTP Server: mod_proxy_ajp:
     Heap Buffer Over-Read Due to Missing Null-Termination Check
     (ajp_msg_get_string) (cve.mitre.org)
     Improper Null Termination, Out-of-bounds Read vulnerability in
     Apache HTTP Server.
     This issue affects Apache HTTP Server: through 2.4.66.
     Credits: Tianshuo Han (<hantianshuo233@gmail.com>)

  *) SECURITY: CVE-2026-33857: Apache HTTP Server: Off-by-one OOB
     reads in AJP getter functions (cve.mitre.org)
     Out-of-bounds Read vulnerability in mod_proxy_ajp of
     Apache HTTP Server.
     This issue affects Apache HTTP Server: through 2.4.66.
     Credits: Elhanan Haenel

  *) SECURITY: CVE-2026-33523: Apache HTTP Server: multiple modules:
     HTTP response splitting forwarding malicious status line
     (cve.mitre.org)
     HTTP response splitting vulnerability in multiple Apache HTTP
     Server modules with untrusted or compromised backend servers.
     This issue affects Apache HTTP Server: from through 2.4.66.
     Credits: Haruki Oyama (Waseda University)

  *) SECURITY: CVE-2026-33007: Apache HTTP Server: mod_authn_socache
     crash (cve.mitre.org)
     A NULL pointer dereference in the mod_authn_socache in Apache
     HTTP Server 2.4.66 and earlier allows an unauthenticated remote
     user to crash a child process in a caching forward proxy
     configuration.
     Credits: Pavel Kohout, Aisle Research, Aisle.com

  *) SECURITY: CVE-2026-33006: Apache HTTP Server: mod_auth_digest
     timing attack (cve.mitre.org)
     A timing attack against mod_auth_digest in Apache HTTP Server
     2.4.66 allows a bypass of Digest authentication by a remote
     attacker.
     Credits: Nitescu Lucian

  *) SECURITY: CVE-2026-29169: Apache HTTP Server: mod_dav_lock
     indirect lock crash (cve.mitre.org)
     A NULL pointer dereference in mod_dav_lock in Apache HTTP Server
     2.4.66 and earlier may allow an attacker to crash the server
     with a malicious request.mod_dav_lock is not used internally by
     mod_dav or mod_dav_fs.
     The only known use-case for mod_dav_lock was mod_dav_svn from
     Apache Subversion earlier than version 1.2.0.
     Users are recommended to upgrade to version 2.4.67, which fixes
     this issue, or remove mod_dav_lock.
     Credits: Pavel Kohout, Aisle Research, Aisle.com

  *) SECURITY: CVE-2026-29168: Apache HTTP Server: mod_md
     unrestricted OCSP response (cve.mitre.org)
     Allocation of Resources Without Limits or Throttling
     vulnerability in Apache HTTP Server's  mod_md via OCSP response
     data.
     This issue affects Apache HTTP Server: from 2.4.30 through
     2.4.66.
     Credits: Pavel Kohout, Aisle Research, Aisle.com

  *) SECURITY: CVE-2026-28780: Apache HTTP Server: buffer overflow in
     mod_proxy_ajp via  ajp_msg_check_header() (cve.mitre.org)
     Heap-based Buffer Overflow vulnerability in mod_proxy_ajp of
     Apache HTTP Server.
     If mod_proxy_ajp connects to a malicious AJP server this AJP
     server can send a malicious AJP message back to mod_proxy_ajp
     and cause it to write 4 attacker controlled bytes after the end
     of a heap based buffer.
     This issue affects Apache HTTP Server: through 2.4.66.
     Credits: Andrew Lacambra

  *) SECURITY: CVE-2026-24072: Apache HTTP Server: mod_rewrite
     elevation of privileges via ap_expr (cve.mitre.org)
     An escalation of privilege bug in various modules in Apache HTTP
     2.4.66 and earlier allows local .htaccess authors to read files
     with the privileges of the httpd user.
     Credits: y7syeu

  *) SECURITY: CVE-2026-23918: Apache HTTP Server: http2: double free
     and possible RCE on early reset (cve.mitre.org)
     Double Free and possible RCE vulnerability in Apache HTTP Server
     with the HTTP/2 protocol.
     This issue affects Apache HTTP Server: 2.4.66.
     Credits: Bartlomiej Dmitruk, striga.ai

  *) mod_md: update to version 2.6.10
     - Fix issue #420 <https://github.com/icing/mod_md/issues/420> by ignoring
       job.json files that claim to have completely finished a certificate
       renewal, but have not produced the necessary result files.

  *) mod_http2: update to version 2.0.39
     Remove streams own memory allocator after reports of memory problems
     with third party modules.
     [Stefan Eissing]

  *) mod_http2: update to version 2.0.38
     Source sync with mod_h2 github repository. No functional change.
     [Stefan Eissing]

  *) Updated conf/mime.types: added vnd.sqlite3, HEIC, HEIF
     [Alexandru Mărășteanu <hello alexei.ro>]

  *) mod_md: update to version 2.6.7
     - Fix a regression in `MDStapleOthers` which broke in v2.6.0 and no longer
       applied, no matter the configuration.

  *) mod_md: update to version 2.6.9
     - Pebble 2.9+ reports another error when terms of service agreement is
       not set. Treating all "userActionRequired" errors as permanent now.

  *) mod_md: update to version 2.6.8
     - Fix the ARI related `replaces` property in ACME order creation to only
       be used when the CA supports ARI and it is enabled in the menu config.
     - Fix compatibility with APR versions before 1.6.0 which do not have
       `apr_cstr_casecmp` and should use `apr_strnatcasecmp` instead.

  *) mod_http2: update to version 2.0.37
     Prevent double purge of a stream, resulting in a double free.
     Fixes PR 69899.
     [Stefan Eissing]

  *) mod_md: Use correct function name when compiling against APR < 1.6.0.
     PR 69954 [Tần Quảng <baobaoxich@gmail.com>]


support/httxt2dbm.c
/* Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * httxt2dbm.c: simple program for converting RewriteMap text files to DBM
 * Rewrite databases for the Apache HTTP server
 *
 */

#include "apr.h"
#include "apr_lib.h"
#include "apr_strings.h"
#include "apr_file_io.h"
#include "apr_file_info.h"
#include "apr_pools.h"
#include "apr_getopt.h"
#include "apu.h"
#include "apr_dbm.h"

#if APR_HAVE_STDLIB_H
#include <stdlib.h> /* for atexit() */
#endif

static const char *input;
static const char *output;
static const char *format;
static const char *shortname;
static apr_file_t *errfile;
static int verbose;

/* From mod_rewrite.c */
#ifndef REWRITE_MAX_TXT_MAP_LINE
#define REWRITE_MAX_TXT_MAP_LINE 1024
#endif

#define NL APR_EOL_STR

#define AVAIL "available"
#define UNAVAIL "unavailable"

static void usage(void)
{
    const char *have_sdbm;
    const char *have_gdbm;
    const char *have_ndbm;
    const char *have_db;

#if APU_HAVE_SDBM
    have_sdbm = AVAIL;
#else
    have_sdbm = UNAVAIL;
#endif
#if APU_HAVE_GDBM
    have_gdbm = AVAIL;
#else
    have_gdbm = UNAVAIL;
#endif
#if APU_HAVE_NDBM
    have_ndbm = AVAIL;
#else
    have_ndbm = UNAVAIL;
#endif
#if APU_HAVE_DB
    have_db = AVAIL;
#else
    have_db = UNAVAIL;
#endif

    apr_file_printf(errfile,
    "%s -- Program to Create DBM Files for use by RewriteMap" NL
    "Usage: %s [-v] [-f format] -i SOURCE_TXT -o OUTPUT_DBM" NL
    NL
    "Options: " NL
    " -v    More verbose output" NL
    NL
    " -i    Source Text File. If '-', use stdin." NL
    NL
    " -o    Output DBM." NL
    NL
    " -f    DBM Format.  If not specified, will use the APR Default." NL
    "           GDBM for GDBM files (%s)" NL
    "           SDBM for SDBM files (%s)" NL
    "           DB   for berkeley DB files (%s)" NL
    "           NDBM for NDBM files (%s)" NL
    "           default for the default DBM type" NL
    NL,
    shortname,
    shortname,
    have_gdbm,
    have_sdbm,
    have_db,
    have_ndbm);
}


static apr_status_t to_dbm(apr_dbm_t *dbm, apr_file_t *fp, apr_pool_t *pool)
{
    apr_status_t rv = APR_SUCCESS;
    char line[REWRITE_MAX_TXT_MAP_LINE + 1]; /* +1 for \0 */
    apr_datum_t dbmkey;
    apr_datum_t dbmval;
    apr_pool_t* p;

    apr_pool_create(&p, pool);

    while (apr_file_gets(line, sizeof(line), fp) == APR_SUCCESS) {
        char *c, *value;

        if (*line == '#' || apr_isspace(*line)) {
            continue;
        }

        c = line;

        while (*c && !apr_isspace(*c)) {
            ++c;
        }

        if (!*c) {
            /* no value. solid line of data. */
            continue;
        }

        dbmkey.dptr = apr_pstrmemdup(p, line,  c - line);
        dbmkey.dsize = (c - line);

        while (apr_isspace(*c)) {
            ++c;
        }

        if (!*c) {
            apr_pool_clear(p);
            continue;
        }

        value = c;

        while (*c && !apr_isspace(*c)) {
            ++c;
        }

        dbmval.dptr = apr_pstrmemdup(p, value,  c - value);
        dbmval.dsize = (c - value);

        if (verbose) {
            apr_file_printf(errfile, "    '%s' -> '%s'" NL,
                            dbmkey.dptr, dbmval.dptr);
        }

        rv = apr_dbm_store(dbm, dbmkey, dbmval);

        apr_pool_clear(p);

        if (rv != APR_SUCCESS) {
            break;
        }
    }

    return rv;
}

int main(int argc, const char *const argv[])
{
    apr_pool_t *pool;
    apr_status_t rv = APR_SUCCESS;
    apr_getopt_t *opt;
    const char *opt_arg;
    char ch;
    apr_file_t *infile;
    apr_dbm_t *outdbm;

    apr_app_initialize(&argc, &argv, NULL);
    atexit(apr_terminate);

    verbose = 0;
    format = NULL;
    input = NULL;
    output = NULL;

    apr_pool_create(&pool, NULL);

    if (argc) {
        shortname = apr_filepath_name_get(argv[0]);
    }
    else {
        shortname = "httxt2dbm";
    }

    apr_file_open_stderr(&errfile, pool);
    rv = apr_getopt_init(&opt, pool, argc, argv);

    if (rv != APR_SUCCESS) {
        apr_file_printf(errfile, "Error: apr_getopt_init failed." NL NL);
        return 1;
    }

    if (argc <= 1) {
        usage();
        return 1;
    }

    while ((rv = apr_getopt(opt, "vf::i::o::", &ch, &opt_arg)) == APR_SUCCESS) {
        switch (ch) {
        case 'v':
            if (verbose) {
                apr_file_printf(errfile, "Error: -v can only be passed once" NL NL);
                usage();
                return 1;
            }
            verbose = 1;
            break;
        case 'f':
            if (format) {
                apr_file_printf(errfile, "Error: -f can only be passed once" NL NL);
                usage();
                return 1;
            }
            format = apr_pstrdup(pool, opt_arg);
            break;
        case 'i':
            if (input) {
                apr_file_printf(errfile, "Error: -i can only be passed once" NL NL);
                usage();
                return 1;
            }
            input = apr_pstrdup(pool, opt_arg);
            break;
        case 'o':
            if (output) {
                apr_file_printf(errfile, "Error: -o can only be passed once" NL NL);
                usage();
                return 1;
            }
            output = apr_pstrdup(pool, opt_arg);
            break;
        }
    }

    if (rv != APR_EOF) {
        apr_file_printf(errfile, "Error: Parsing Arguments Failed" NL NL);
        usage();
        return 1;
    }

    if (!input) {
        apr_file_printf(errfile, "Error: No input file specified." NL NL);
        usage();
        return 1;
    }

    if (!output) {
        apr_file_printf(errfile, "Error: No output DBM specified." NL NL);
        usage();
        return 1;
    }

    if (!format) {
        format = "default";
    }

    if (verbose) {
        apr_file_printf(errfile, "DBM Format: %s" NL, format);
    }

    if (!strcmp(input, "-")) {
        rv = apr_file_open_stdin(&infile, pool);
    }
    else {
        rv = apr_file_open(&infile, input, APR_READ|APR_BUFFERED,
                           APR_OS_DEFAULT, pool);
    }

    if (rv != APR_SUCCESS) {
        apr_file_printf(errfile,
                        "Error: Cannot open input file '%s': (%d) %pm" NL NL,
                         input, rv, &rv);
        return 1;
    }

    if (verbose) {
        apr_file_printf(errfile, "Input File: %s" NL, input);
    }

    rv = apr_dbm_open_ex(&outdbm, format, output, APR_DBM_RWCREATE,
                    APR_OS_DEFAULT, pool);

    if (APR_STATUS_IS_ENOTIMPL(rv)) {
        apr_file_printf(errfile,
                        "Error: The requested DBM Format '%s' is not available." NL NL,
                         format);
        return 1;
    }

    if (rv != APR_SUCCESS) {
        apr_file_printf(errfile,
                        "Error: Cannot open output DBM '%s': (%d) %pm" NL NL,
                         output, rv, &rv);
        return 1;
    }

    if (verbose) {
        apr_file_printf(errfile, "DBM File: %s" NL, output);
    }

    rv = to_dbm(outdbm, infile, pool);

    if (rv != APR_SUCCESS) {
        apr_file_printf(errfile,
                        "Error: Converting to DBM: (%d) %pm" NL NL,
                         rv, &rv);
        return 1;
    }

    apr_dbm_close(outdbm);

    if (verbose) {
        apr_file_printf(errfile, "Conversion Complete." NL);
    }

    return 0;
}

